File: Easy/SameTree.cpp
/*
  100. Same Tree

  Given the roots of two binary trees p and q, write a function to check if they
  are the same or not. Two binary trees are considered the same if they are
  structurally identical, and the nodes have the same value.

  Example:
  Input: p = [1,2,3], q = [1,2,3]
  Output: true
 */

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right){} *
};

class Solution {
public:
  // Utilises a breadth-first search to iterate through each tree simultaneously
  // and ensure they match. If a mismatch occurs at any point, return false and
  // the overall return will be false. Time Complexity: O(n).
  bool BFS_Search(TreeNode *same, TreeNode *other) {
    if (same == nullptr && other == nullptr) {
      return true;
    }
    if (same == nullptr || other == nullptr || same->val != other->val) {
      return false;
    }
    return BFS_Search(same->left, other->left) &&
           BFS_Search(same->right, other->right);
  }

  bool isSameTree(TreeNode *p, TreeNode *q) { return BFS_Search(p, q); }
};

File: Easy/SameTree.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution(object):
    def isSameTree(self, p, q):
        """
        Given the roots of two binary trees p and q, write a
        function to check if they are the same or not. Two
        binary trees are considered the same if they are
        structurally identical, and the nodes have the same value.

        :type p: TreeNode
        :type q: TreeNode
        :rtype: bool
        """

        def dfs(p, q):
            # If neither node exists (is NULL)
            if not p and not q:
                return True
            # If either node does not exist but the other does
            elif not p or not q:
                return False
            # If there is a node p,q that exists
            else:
                # If the value of p != q
                if p.val != q.val:
                    return False
                else:
                    left = dfs(p.left, q.left)
                    right = dfs(p.right, q.right)
            if left and right:
                return True
            else:
                return False

        return dfs(p, q)
